src/coreclr/jit/morph.cpp

```diff
@@ -9381,185 +9428,546 @@ GenTree* Compiler::fgOptimizeHWIntrinsic(GenTreeHWIntrinsic* node)
+                                                              reverseCond);
+                NamedIntrinsic newId =
+                    GenTreeHWIntrinsic::GetHWIntrinsicIdForCmpOp(this, op1Oper, lookupType, cmpOp1, cmpOp2,
+                                                                 op1SimdBaseType, op1SimdSize, op1IsScalar,
+                                                                 reverseCond);
+
+                if (newId != NI_Illegal)
                 {
-                    GenTreeHWIntrinsic* hw             = op2->AsHWIntrinsic();
-                    genTreeOps          hwOper         = hw->GetOperForHWIntrinsicId(&isScalar);
-                    var_types           hwSimdBaseType = hw->GetSimdBaseType();
+                    op1Intrin->ResetHWIntrinsicId(newId, cmpOp1, cmpOp2);
+                    ExtractEffectiveOp(GT_NOT, node, /* destroyNodes */ true);
 
-                    if (isScalar)
+                    if (lookupType != op1RetType)
                     {
-                        return node;
-                    }
+                        assert(varTypeIsSIMD(op1RetType));
+                        assert(varTypeIsMask(lookupType));
 
-                    if (varTypeToSigned(simdBaseType) != varTypeToSigned(hwSimdBaseType))
-                    {
-                        // We need the base types to be of the same kind and size
-                        // that is, we can't mix floating-point and integers or int and long
-                        // but we can mix int and uint or long and ulong.
-                        return node;
+                        op1Intrin->gtType = lookupType;
+                        op1Intrin = gtNewSimdCvtMaskToVectorNode(retType, op1Intrin, op1SimdBaseJitType, op1SimdSize)
+                                        ->AsHWIntrinsic();
                     }
 
-                    if (hwOper == GT_SUB)
+                    if (cvtIntrin != nullptr)
                     {
-                        if (varTypeIsIntegral(hwSimdBaseType) && hw->Op(1)->IsVectorZero())
-                        {
-                            GenTree* result = hw->Op(2);
-                            DEBUG_DESTROY_NODE(hw->Op(1));
-                            DEBUG_DESTROY_NODE(hw);
-                            DEBUG_DESTROY_NODE(op1);
-                            DEBUG_DESTROY_NODE(node);
-                            return result;
-                        }
+                        DEBUG_DESTROY_NODE(cvtIntrin);
                     }
+                    return fgMorphHWIntrinsicRequired(op1Intrin);
                 }
-#endif // !TARGET_ARM64 && !TARGET_XARCH
             }
             break;
         }
 
         // Transforms:
-        // 1. (v1 ^ AllBitsSet) to VectorXxx.OnesComplement(v1)
-        // 2. (v1 ^ -0.0) to VectorXxx.Negate(v1); for floating-point
-        case GT_XOR:
+        // 1. (-v1) - (-v2) to v2 - v1
```

Comment:
comment not related to this PR in particular. But I wish we had infrastructure to pattern match and replace this in morph....currently we are just adding this manually and making it more bug prone.